From 54399ee441d922d89c32909e2028f899f6091cd6 Mon Sep 17 00:00:00 2001
From: Niels Dossche <7771979+nielsdos@users.noreply.github.com>
Date: Sun, 7 Jan 2024 21:59:28 +0100
Subject: [PATCH 6/6] Patch out unused CSS style code

---
 source/lexbor/css/rule.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/source/lexbor/css/rule.h b/source/lexbor/css/rule.h
index 308dced..d192a01 100644
--- a/source/lexbor/css/rule.h
+++ b/source/lexbor/css/rule.h
@@ -361,6 +361,7 @@ lxb_css_rule_ref_dec(lxb_css_rule_t *rule)
 lxb_inline void
 lxb_css_rule_ref_dec_destroy(lxb_css_rule_t *rule)
 {
+#if 0
     if (rule->ref_count > 0) {
         rule->ref_count--;
     }
@@ -368,6 +369,7 @@ lxb_css_rule_ref_dec_destroy(lxb_css_rule_t *rule)
     if (rule->ref_count == 0) {
         (void) lxb_css_rule_destroy(rule, true);
     }
+#endif
 }
 
 lxb_inline void
-- 
2.51.2

