diff -ur Magdir.orig/rpm Magdir/rpm
--- Magdir.orig/rpm	2025-02-09 12:44:15.956602145 +0100
+++ Magdir/rpm	2025-02-09 12:45:16.560624196 +0100
@@ -34,6 +34,7 @@
 >>8	beshort		22		RISC-V64
 >>8	beshort		23		LoongArch64
 >>8	beshort		255		noarch
+>>10	string		x		%s
 
 #delta RPM    Daniel Novotny (dnovotny@redhat.com)
 0	string		drpm		Delta RPM
diff -ur Magdir.orig/securitycerts Magdir/securitycerts
--- Magdir.orig/securitycerts	2025-02-09 12:44:15.956628770 +0100
+++ Magdir/securitycerts	2025-02-09 12:44:44.077144142 +0100
@@ -4,3 +4,5 @@
 0	search/1		-----BEGIN\ CERTIFICATE------	RFC1421 Security Certificate text
 0	search/1		-----BEGIN\ NEW\ CERTIFICATE	RFC1421 Security Certificate Signing Request text
 0	belong	0xedfeedfe	Sun 'jks' Java Keystore File data
+
+0	string \0volume_key	volume_key escrow packet
