--- re.c.ruby_orig	Tue Feb  4 15:52:29 2003
+++ re.c	Tue Mar 18 19:37:49 2003
@@ -380,7 +380,8 @@ make_regexp(s, len, flag)
     int len, flag;
 {
     Regexp *rp;
-    char *err;
+    char err[REG_MAX_ERROR_MESSAGE_LEN];
+    int r;
 
     /* Handle escaped characters first. */
 
@@ -389,16 +390,17 @@ make_regexp(s, len, flag)
        from that.
     */
 
-    rp = ALLOC(Regexp);
-    MEMZERO((char *)rp, Regexp, 1);
-    rp->buffer = ALLOC_N(char, 16);
-    rp->allocated = 16;
-    rp->fastmap = ALLOC_N(char, 256);
+    r = re_alloc_pattern(&rp);
+    if (r) {
+      re_error_code_to_str(err, r);
+      rb_reg_raise(s, len, err, 0);
+    }
+      
     if (flag) {
 	rp->options = flag;
     }
-    err = re_compile_pattern(s, len, rp);
-    if (err != NULL) {
+    r = re_compile_pattern(s, len, rp, err);
+    if (r != 0) {
 	rb_reg_raise(s, len, err, 0);
     }
 
@@ -532,14 +534,14 @@ rb_reg_prepare_re(re)
     }
 
     if (need_recompile) {
-	char *err;
+	char err[REG_MAX_ERROR_MESSAGE_LEN];
+	int r;
 
 	if (FL_TEST(re, KCODE_FIXED))
 	    kcode_set_option(re);
 	rb_reg_check(re);
-	RREGEXP(re)->ptr->fastmap_accurate = 0;
-	err = re_compile_pattern(RREGEXP(re)->str, RREGEXP(re)->len, RREGEXP(re)->ptr);
-	if (err != NULL) {
+	r = re_recompile_pattern(RREGEXP(re)->str, RREGEXP(re)->len, RREGEXP(re)->ptr, err);
+	if (r != 0) {
 	    rb_reg_raise(RREGEXP(re)->str, RREGEXP(re)->len, err, re);
 	}
     }
