--- sqlite3.c.orig	2011-04-17 10:54:01.000000000 -0700
+++ sqlite3.c	2011-05-12 22:57:02.000000000 -0700
@@ -1,3 +1,7 @@
+#if defined(_MSC_VER) && _MSC_VER < 1300
+#pragma optimize("", off)
+#endif
+
 /******************************************************************************
 ** This file is an amalgamation of many separate C source files from SQLite
 ** version 3.7.6.2.  By combining all the individual C code files into this 
@@ -125956,3 +125960,7 @@
 #endif /* !defined(SQLITE_CORE) || defined(SQLITE_ENABLE_FTS3) */
 
 /************** End of fts3_icu.c ********************************************/
+
+#if defined(_MSC_VER) && _MSC_VER < 1300
+#pragma optimize("", on)
+#endif
