*** sapi_apache.c.orig	Fri Oct 13 23:12:44 2000
--- sapi_apache.c	Fri Oct 13 23:25:39 2000
***************
*** 69,91 ****
  	/* sending a file handle to another dll is not working
  	// so let zend open it. 
  	*/
! 	file_handle.type = ZEND_HANDLE_FILENAME;
! 	file_handle.handle.fd = 0;
! 	file_handle.filename = SG(request_info).path_translated;
! 	file_handle.opened_path = NULL;
! 	file_handle.free_filename = 0;
! 
  	if (display_source_mode) {
  		zend_syntax_highlighter_ini syntax_highlighter_ini;
  
! 		if (open_file_for_scanning(&file_handle CLS_CC)==SUCCESS) {
! 			php_get_highlight_struct(&syntax_highlighter_ini);
! 			zend_highlight(&syntax_highlighter_ini);
  			return OK;
  		} else {
  			return NOT_FOUND;
  		}
  	} else {
  		(void) php_execute_script(&file_handle CLS_CC ELS_CC PLS_CC);
  	}
  	
--- 69,91 ----
  	/* sending a file handle to another dll is not working
  	// so let zend open it. 
  	*/
! 	
  	if (display_source_mode) {
  		zend_syntax_highlighter_ini syntax_highlighter_ini;
  
! 		php_get_highlight_struct(&syntax_highlighter_ini);
! 		if (highlight_file(SG(request_info).path_translated,&syntax_highlighter_ini)){
  			return OK;
  		} else {
  			return NOT_FOUND;
  		}
  	} else {
+ 		file_handle.type = ZEND_HANDLE_FILENAME;
+ 		file_handle.handle.fd = 0;
+ 		file_handle.filename = SG(request_info).path_translated;
+ 		file_handle.opened_path = NULL;
+ 		file_handle.free_filename = 0;
+ 
  		(void) php_execute_script(&file_handle CLS_CC ELS_CC PLS_CC);
  	}
  	
